### PR TITLE
Load all dalvik files in an APK

### PR DESCRIPTION
**Pull request by wargio**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

This loads all the dex files in an APK

### CODE REVIEW

File: librz/core/cfile.c [hunk 1]
```diff
@@ -892,32 +892,58 @@ typedef struct {
 	RzBin *bin;
 } RzCoreLinkData;
 
-static bool linkcb(void *user, void *data, ut32 id) {
-	RzCoreLinkData *ld = user;
-	RzIODesc *desc = (RzIODesc *)data;
-
+static bool resolve_import_cb(RzCoreLinkData *ld, RzIODesc *desc, ut32 id) {
```
Comment (XVilka) [inline librz/core/cfile.c:895]: Please also add `rz_return_val_if_fail(ld && desc, false)`.

File: librz/core/cfile.c [hunk 2]
```diff
@@ -892,32 +892,58 @@ typedef struct {
...
+	rz_list_foreach (symbols, iter, sym) {
+		if (!strcmp(sym->name, ld->name)) {
+			ld->addr = sym->vaddr;
+			return false;
 		}
 	}
 	return true;
 }
 
+static bool map_multi_dex(RzCore *core, RzIODesc *desc, ut32 id) {
```
Comment (XVilka) [inline librz/core/cfile.c:913]: Same here.

File: librz/core/cfile.c [hunk 3]
```diff
@@ -892,32 +892,58 @@ typedef struct {
...
+	if (!rz_str_endswith(desc->name, ".dex")) {
+		return true;
+	}
+
+	ut64 size = rz_io_desc_size(desc);
+	ut64 baddr = rz_io_map_next_available(core->io, 0x200000, size, 0x100000);
```
Comment (XVilka) [inline librz/core/cfile.c:920]: What are these hardcoded values?

File: librz/core/cfile.c [hunk 4]
```diff
@@ -892,32 +892,58 @@ typedef struct {
...
+		opt.xtr_idx = 0;
+		RzBinFile *binfile = rz_bin_open_io(core->bin, &opt);
+		if (!binfile) {
+			RZ_LOG_ERROR("Cannot load bin file %s.\n", desc->name);
+			return true;
+		}
+
+		rz_pvector_push(&cf->binfiles, binfile);
+		rz_core_bin_apply_all_info(core, binfile);
+		rz_core_cmd0(core, "\"(fix-dex,wx `ph sha1 $s-32 @32` @12 ; wx `ph adler32 $s-12 @12` @8)\"\n");
```
Comment (XVilka) [inline librz/core/cfile.c:936]: At least some of that command could be converted into the API - `ph` and `wx` for sure.

File: librz/core/cfile.c [hunk 5]
```diff
@@ -892,32 +892,58 @@ typedef struct {
...
+	}
+
+	return true;
+}
+
 RZ_API bool rz_core_bin_load(RzCore *r, const char *filenameuri, ut64 baddr) {
```
Comment (XVilka) [inline librz/core/cfile.c]: Please add doxygen, also mark `filenameuri` as `RZ_NULLABLE`

File: librz/core/cfile.c [hunk 6]
```diff
@@ -1027,44 +1057,48 @@ RZ_API bool rz_core_bin_load(RzCore *r, const char *filenameuri, ut64 baddr) {
...
 			ut64 baddr = rz_io_map_location(r->io, 0x200000);
 			if (baddr != UT64_MAX) {
 				rz_core_file_loadlib(r, lib, baddr);
 			}
 		}
-		rz_core_cmd0(r, "obb 0;s entry0");
-		rz_config_set_i(r->config, "bin.at", true);
-		eprintf("[bin.libs] Linking imports...\n");
+
+		rz_core_cmd0(r, "ob 0; s entry0");
```
Comment (XVilka) [inline librz/core/cfile.c:1077]: Please use the API

File: librz/core/cfile.c [hunk 7]
```diff
@@ -1115,16 +1149,16 @@ RZ_API RzCoreFile *rz_core_file_open_many(RzCore *r, const char *file, int perm,
...
-			rz_core_bin_load(r, fd->name, loadaddr);
+			rz_core_bin_load(r, desc->name, loadaddr);
 		}
 	}
 	return NULL;
 }
 
-/* loadaddr is r2 -m (mapaddr) */
+/* loadaddr is rizin -m (mapaddr) */
 RZ_API RzCoreFile *rz_core_file_open(RzCore *r, const char *file, int flags, ut64 loadaddr) {
```
Comment (XVilka) [inline librz/core/cfile.c:1164]: Please add `RZ_NONNULL` for `file`. Ideally, doxygen as well, while you are on it.

File: librz/io/io_map.c [hunk 8]
```diff
@@ -421,13 +418,9 @@ RZ_API bool rz_io_map_resize(RzIO *io, ut32 id, ut64 newsize) {
 }
 
 // find a location that can hold enough bytes without overlapping
```
Comment (XVilka) [inline librz/io/io_map.c]: Could you please convert this description into Doxygen one?

File: librz/io/p/io_zip.c [hunk 9]
```diff
@@ -157,7 +157,7 @@ static int rz_io_zip_slurp_file(RzIOZipFileObj *zfo) {
 	return res;
 }
 
-RzList *rz_io_zip_get_files(char *archivename, ut32 perm, int mode, int rw) {
+RzList *rz_io_zip_get_files(const char *archivename, ut32 perm, int mode, int rw) {
```
Comment (XVilka) [inline librz/io/p/io_zip.c:160]: Can be `archivename` `NULL` here? Please add a check then. Please also fix the indentation after, and convert pattern of `if (zipArch)` into `if (!zipArch)`.

File: librz/io/p/io_zip.c [hunk 10]
```diff
@@ -370,9 +370,83 @@ char *rz_io_zip_get_by_file_idx(const char *archivename, const char *idx, ut32 p
 	return filename;
 }
 
+static char *find_ipa_binary(const char *filename, int rw, int mode) {
+	RzList *files = NULL;
```
Comment (XVilka) [inline librz/io/p/io_zip.c:377]: Check for filename here - could it be NULL or not?

File: librz/io/p/io_zip.c [hunk 11]
```diff
@@ -370,9 +370,83 @@ char *rz_io_zip_get_by_file_idx(const char *archivename, const char *idx, ut32 p
...
+		}
+		RZ_FREE(zip_filename);
+	}
+	rz_list_free(files);
+
+	return zip_filename;
+}
+
+static char *find_apk_binary(const char *filename, int rw, int mode, RzIO *io) {
+	RzList *files = NULL;
```
Comment (XVilka) [inline librz/io/p/io_zip.c:410]: Same here

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/core/cfile.c (hunk 3 above)
Comment (wargio) [reply librz/core/cfile.c:920]: is the base address and the advance value
so if it fails `base + advance` till it finds the next empty address space

File: librz/core/cfile.c (hunk 4 above)
Comment (wargio) [reply librz/core/cfile.c:936]: tbh, the whole cfile needs to be rewritten. i would keep this like this for now

Comment (wargio) [reply librz/core/cfile.c:936]: i will rewrite the whole this later

File: librz/core/cfile.c (hunk 5 above)
Comment (wargio) [reply librz/core/cfile.c]: no need to add doxygen. this whole core thing needs to be rewritten

File: librz/core/cfile.c (hunk 6 above)
Comment (wargio) [reply librz/core/cfile.c:1077]: same here. the whole thing needs to be rewritten. i would avoid changing this for now.

File: librz/io/p/io_zip.c (hunk 10 above)
Comment (wargio) [reply librz/io/p/io_zip.c:377]: no it can't

File: librz/io/p/io_zip.c (hunk 11 above)
Comment (wargio) [reply librz/io/p/io_zip.c:410]: see other comment

File: librz/io/p/io_zip.c (hunk 9 above)
Comment (wargio) [reply librz/io/p/io_zip.c:160]: no it can't.

**Review by XVilka — APPROVED** [review]

**Pull request merged**